src/v/ssx/include/ssx/sharded_ptr.h

```diff
@@ -84,6 +85,20 @@ class sharded_ptr {
         return reset(std::make_shared<T>(std::forward<Args>(args)...));
     }
 
+    /// updates the managed object u with an update function
+    /// The update function takes a copy of the underlying object and creates a
+    /// new object using the update function. The update is executed inside the
+    /// guard of the update mutex to ensure that updates are executed
+    /// sequentially.
+    /// Must be called on the home shard and is safe to call concurrently.
+    /// returns an ss::broken_semaphore if stop() has been called.
+    template<typename F>
+    requires std::invocable<F, T> && std::same_as<T, std::invoke_result_t<F, T>>
+    ss::future<> update(F&& update_func) {
+        co_return co_await reset(
+          update_func(local() ? std::as_const(*local()) : T{}));
```

Comment:
Updated. I've also now realised that I had to switch up update and reset such that reset is implemented in terms of the update and not the other way around, because with `reset(update_func(local() ? std::as_const(*local()) : T{}))` the update happens outside of the mutex.

Reply:
> Updated. I've also now realised that I had to switch up update and reset such that reset is implemented in terms of the update and not the other way around, because with `reset(update_func(local() ? std::as_const(*local()) : T{}))` the update happens outside of the mutex.

The `update_func` is called outside of the mutex, but that's ok, since it's just an observer, it's either taking a newly constructed value or a const&, so no changes can be made to the underlying data structure.

The actual shard updates are done with the returned value passed to reset, which takes the mutex.